### DISCUSSION TITLE
Execute custom function before navigating; add parameters to Route

### DISCUSSION THREAD

**Discussion by matthew-nm** [Milad-Akarie/auto_route_library#1984] [General]
Hi, I have a page in my app that works best if the state (via riverpod) is set before navigating. This code is synchronous. Currently, I have the logic execute in my button call just before calling the router, but I feel like this method is error-prone because I could easily forget to call the state logic.

```dart
onTap: () {
  ref.read(noteFormNotifierProvider.notifier).initialized(note);
  ref.read(formTodosNotifierProvider.notifier).initialized(note.todos);
  context.router.push(NoteFormRoute(user: user));
},
```

Is it possible to:
1. Have the NoteFormRoute execute a custom function before being pushed so I can put the two state inits in there?
2. Have the NoteFormRoute take in extra named parameters (that won't go into the page constructor) so I can pass in the `ref` and `note` objects for the custom function to use?

If so, I would then call the route as,
 ```dart
onTap: () =>
  context.router.push(NoteFormRoute(user: user, ref: ref, note: note));
```

My first thought was to simply do the state initialization within the NoteFormPage, but you aren't allowed to do `ref.read(....notifer).func(...)` within a widget's build cycle. My next thought was to use a `FutureBuilder` or riverpod `AsyncValue`, but both of these are async resolvers that I can't quite get to work right but also would draw some loading frames first which is unnecessary and feels janky given that the code I'm executing is synchronous (I already have the data from the previous page, just need to set the states).

Thanks for any help or suggestions!

**Comment by Milad-Akarie** [comment]
@matthew-nm have you tried doing your login inside of a route guard?

**Reply by matthew-nm** [reply]
Is it possible for the guards to access the page parameters? It looks like the guards accept parameters when they are initialized, but I was hoping to have a function perform an action on the parameters within `context.router.push(NoteFormRoute(user: user, ref: ref, note: note));`.

The guard gets access to 'resolver' and 'router', but I don't believe either of these will let me access the params 'ref' and 'note'.

**Comment by Milad-Akarie** [comment]
@matthew-nm I highly recommend that you pass the ref to the Guard's constructor, and yes you can access the args of the guarded route using `resolver.route.args as NoteFormArgs`

**Reply by matthew-nm** [reply]
Excellent! This works perfectly.

For those interesting in following along, I typed (set the type) the parameters in the page's constructor, then typed them in the guard via `final args = resolver.route.args as NoteFormRouteArgs;`.
